crates/table/src/table.rs

```diff
@@ -260,19 +261,122 @@ pub enum ReadViaBsatnError {
     DecodeError(#[from] DecodeError),
 }
 
+#[derive(Error, Debug)]
+#[error("Cannot change the columns of table `{table_name}` with id {table_id} from `{old:?}` to `{new:?}`")]
+pub struct ChangeColumnsError {
+    table_id: TableId,
+    table_name: Box<str>,
+    old: Vec<ColumnSchema>,
+    new: Vec<ColumnSchema>,
+}
+
+/// Computes the parts of a table definition, that are row type dependent, from the row type.
+fn table_row_type_dependents(row_type: ProductType) -> (RowTypeLayout, StaticLayoutInTable, VarLenVisitorProgram) {
+    let row_layout: RowTypeLayout = row_type.into();
+    let static_layout = StaticLayout::for_row_type(&row_layout).map(|sl| (sl, static_bsatn_validator(&row_layout)));
+    let visitor_prog = row_type_visitor(&row_layout);
+
+    (row_layout, static_layout, visitor_prog)
+}
+
 // Public API:
 impl Table {
     /// Creates a new empty table with the given `schema` and `squashed_offset`.
     pub fn new(schema: Arc<TableSchema>, squashed_offset: SquashedOffset) -> Self {
-        let row_layout: RowTypeLayout = schema.get_row_type().clone().into();
-        let static_layout = StaticLayout::for_row_type(&row_layout).map(|sl| (sl, static_bsatn_validator(&row_layout)));
-        let visitor_prog = row_type_visitor(&row_layout);
+        let (row_layout, static_layout, visitor_prog) = table_row_type_dependents(schema.get_row_type().clone());
+
         // By default, we start off with an empty pointer map,
         // which is removed when the first unique index is added.
         let pm = Some(PointerMap::default());
         Self::new_raw(schema, row_layout, static_layout, visitor_prog, squashed_offset, pm)
     }
 
+    /// Change the columns of `self` to those in `column_schemas`.
+    ///
+    /// Returns an error if the new list of column is incompatible with the old.
+    ///
+    /// # Safety
+    ///
+    /// If `!validate`, the caller must ensure that `column_schemas`
+    /// forms a row type and layout that is compatible with the rows existing in `self`.
+    pub unsafe fn change_columns_to(
+        &mut self,
+        column_schemas: Vec<ColumnSchema>,
+        validate: bool,
```

Comment:
I think that this `bool` should be instead a enum to make more clear to the callers what `true, false` mean. (or split in 2 methods?)